Sources/eppo/Poller.swift

```diff
@@ -0,0 +1,198 @@
+    var executeCount: Int { get }
+    func schedule(deadline: TimeInterval, callback: @escaping () async -> Void)
+    func cancel()
+}
+
+// Real timer implementation
+@MainActor
+class RealTimer: TimerType {
+    var executeCount: Int = 0
+    private var timer: Timer?
+    
+    init() {
+        executeCount = 0
+    }
+    
+    func schedule(deadline: TimeInterval, callback: @escaping () async -> Void) {
+        executeCount += 1
+        timer?.invalidate()
+        
+        timer = Timer.scheduledTimer(withTimeInterval: deadline, repeats: false) { _ in
+            Task {
+                await callback()
+            }
+        }
+        // Make sure the timer is added to the main run loop
+        RunLoop.main.add(timer!, forMode: .common)
+    }
+    
+    func cancel() {
+        timer?.invalidate()
+        timer = nil
+    }
+}
+
+// Test timer implementation that respects scheduling
+@MainActor
+class TestTimer: TimerType {
+    private(set) var executeCount: Int = 0
+    private var isRunning = true
+    private var startTime: TimeInterval
+    
+    init() {
+        self.startTime = Date().timeIntervalSince1970
+    }
+    
+    func schedule(deadline: TimeInterval, callback: @escaping () async -> Void) {
+        executeCount += 1
+        
+        if isRunning {
+            Task {
+                // Convert deadline to nanoseconds and wait
+                let delayNanos = UInt64(deadline * 1_000_000_000) // Convert seconds to nanoseconds
+                try? await Task.sleep(nanoseconds: delayNanos)
+                
+                if self.isRunning {
+                    await callback()
+                }
+            }
+        }
+    }
+    
+    func cancel() {
+        isRunning = false
+    }
+}
+
+@MainActor
+public class Poller {
+    private let intervalMs: Int
+    private let jitterMs: Int
+    private let callback: () async throws -> Void
+    
+    private var stopped = true
+    private var failedAttempts = 0
+    private var nextPollMs: Int
+    private var logger: PollerLogger
+    private let timer: TimerType
+    
+    public init(
+        intervalMs: Int,
+        jitterMs: Int,
+        callback: @escaping () async throws -> Void
+    ) async {
+        self.intervalMs = intervalMs
+        self.jitterMs = jitterMs
+        self.callback = callback
+        self.nextPollMs = intervalMs
+        self.logger = PollerLogger()
+        self.timer = RealTimer()
+    }
+    
+    // Internal initializer for testing
+    init(
+        intervalMs: Int,
+        jitterMs: Int,
+        callback: @escaping () async throws -> Void,
+        logger: PollerLogger,
+        timer: TimerType
+    ) async {
+        self.intervalMs = intervalMs
+        self.jitterMs = jitterMs
+        self.callback = callback
+        self.nextPollMs = intervalMs
+        self.logger = logger
+        self.timer = timer
+    }
+    
+    public func start() async throws {
+        stopped = false
```

Comment:
Do you need to reset failedAttempts here as well> If the poller is stopped during a retry, then restarted, I think you would want to reset the failed attempt counter.

Reply:
Also a great suggestion! Added

https://github.com/Eppo-exp/eppo-ios-sdk/blob/05122e44edff641fb8d189b4aca6a62bc395bafe/Sources/eppo/Poller.swift#L136-L137